File: 101/03 双指针/04_76.cc
class Solution {
  public:
    string minWindow(string s, string t) {
      vector<int> chars(128, 0);
      vector<bool> flag(128, false);
      for (int i = 0; i < t.size(); ++i) {
        ++chars[t[i]];
        flag[t[i]] = true;
      }
      int count = 0;
      int l = 0;
      int min_l = 0;
      int min_size = s.size() + 1;
      for (int r = 0; r < s.size(); ++r) {
        if (flag[s[r]]) {
          if (--chars[s[r]] >= 0) ++count;
          while (count == t.size()) {
            int current_size = r - l + 1;
            if (current_size < min_size) {
              min_size = current_size;
              min_l = l;
            }
            if (flag[s[l]] && ++chars[s[l]] > 0) --count;
            ++l;
          }
        }
      }
      return min_size > s.size() ? "" : s.substr(min_l, min_size);
    }
};

File: 101/03 双指针/04_76.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function (s, t) {
  const chars = new Array(128).fill(0);
  const flag = new Array(128).fill(false);
  t.split('').forEach((character) => {
    ++chars[character.codePointAt(0)];
    flag[character.codePointAt(0)] = true;
  });
  let count = 0;
  let l = 0;
  let min_l = 0;
  let min_size = s.length + 1;
  for (let r = 0; r < s.length; ++r) {
    if (flag[s[r].codePointAt(0)]) {
      if (--chars[s[r].codePointAt(0)] >= 0) ++count;
      while (count === t.length) {
        const current_size = r - l + 1;
        if (current_size < min_size) {
          min_size = current_size;
          min_l = l;
        }
        if (flag[s[l].codePointAt(0)] && ++chars[s[l].codePointAt(0)] > 0) --count;
        ++l;
      }
    }
  }
  return min_size > s.length ? "" : s.slice(min_l, min_l + min_size);
};
